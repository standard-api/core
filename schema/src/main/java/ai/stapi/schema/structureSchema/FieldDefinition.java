package ai.stapi.schema.structureSchema;

import ai.stapi.serialization.AbstractSerializableObject;
import java.util.Comparator;
import java.util.List;

public class FieldDefinition extends AbstractSerializableObject {

  public static final String SERIALIZATION_TYPE = "FieldDefinition";
  public static final String LIST_STRUCTURE_TYPE = "list";
  public static final String LEAF_STRUCTURE_TYPE = "leaf";
  private String name;
  private Integer min;
  private String max;
  private String description;
  private List<FieldType> types;

  private String parentDefinitionType;

  private FieldDefinition() {
    super(SERIALIZATION_TYPE);
  }

  public FieldDefinition(
      String name,
      Integer min,
      String max,
      String description,
      List<FieldType> types,
      String parentDefinitionType
  ) {
    super(SERIALIZATION_TYPE);
    this.name = name;
    this.min = min;
    this.max = max;
    this.description = description;
    this.types = types.stream()
        .sorted(Comparator.comparing(FieldType::getType))
        .toList();
    this.parentDefinitionType = parentDefinitionType;
  }

  public Integer getMin() {
    return min;
  }

  public String getMax() {
    return max;
  }

  public Float getFloatMax() {
    if (this.max.equals("*")) {
      return Float.POSITIVE_INFINITY;
    }
    return Float.parseFloat(this.max);
  }

  public List<FieldType> getTypes() {
    return types;
  }

  public String getStructureType() {
    return isList() ? LIST_STRUCTURE_TYPE : LEAF_STRUCTURE_TYPE;
  }

  public boolean isList() {
    try {
      var value = Integer.parseInt(this.max);
      return value > 1;
    } catch (NumberFormatException ignored) {

    }
    return this.max.equals("*");
  }

  public boolean isRequired() {
    return this.min > 0;
  }

  public boolean isUnionType() {
    return this.types.size() > 1;
  }

  public boolean isAnyType() {
    return this.types.isEmpty();
  }

  public String getDescription() {
    return description;
  }

  public String getName() {
    return name;
  }

  public String getParentDefinitionType() {
    return parentDefinitionType;
  }
}
