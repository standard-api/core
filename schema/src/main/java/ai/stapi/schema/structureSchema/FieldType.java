package ai.stapi.schema.structureSchema;

import org.apache.commons.lang3.StringUtils;

public class FieldType {

  private final String type;
  private final String originalType;
  private final FieldTypeGroup typeGroup;

  private final boolean isContentReferenced;

  public FieldType(
      String type,
      String originalType
  ) {
    this(type, originalType, FieldTypeGroup.STANDARD);
  }

  public FieldType(
      String type,
      String originalType,
      FieldTypeGroup fieldTypeGroup
  ) {
    this(type, originalType, fieldTypeGroup, false);
  }

  private FieldType(
      String type,
      String originalType,
      FieldTypeGroup fieldTypeGroup,
      boolean isContentReferenced
  ) {
    this.type = type;
    this.originalType = originalType;
    this.typeGroup = fieldTypeGroup;
    this.isContentReferenced = isContentReferenced;
  }

  public static FieldType asReferenceType(String type) {
    return new FieldType(type, "Reference", FieldTypeGroup.REFERENCE);
  }

  public static FieldType asContentReferenced(String type, String originalName) {
    return new FieldType(
        type,
        originalName,
        FieldTypeGroup.ANONYMOUS,
        true
    );
  }

  public static FieldType asAnonymousType(String type, String originalName) {
    if (!originalName.equals("Element") && !originalName.equals("BackboneElement")) {
      throw new RuntimeException(
          "Anonymous type's original name must be either Element or BackboneElement");
    }
    return new FieldType(type, originalName, FieldTypeGroup.ANONYMOUS);
  }

  public static FieldType asBoxedType(String type) {
    String typeName;
    String originalName;
    if (type.startsWith("Boxed")) {
      typeName = type;
      originalName = type.replace("Boxed", "").toLowerCase();
    } else {
      typeName = "Boxed" + StringUtils.capitalize(type);
      originalName = type;
    }
    return new FieldType(typeName, originalName, FieldTypeGroup.BOXED);
  }

  public static FieldType asPlainType(String type) {
    return new FieldType(type, type);
  }

  public String getType() {
    return type;
  }

  public String getOriginalType() {
    return originalType;
  }

  public FieldTypeGroup getTypeGroup() {
    return typeGroup;
  }

  public boolean isReference() {
    return this.originalType.equals("Reference");
  }

  public boolean isAnonymous() {
    return this.originalType.equals("Element") || this.originalType.equals("BackboneElement");
  }

  public boolean isBoxed() {
    return this.type.startsWith("Boxed") && !this.type.equals(this.originalType);
  }

  public boolean isPrimitiveType() {
    return this.type.equals(this.originalType) && Character.isLowerCase(this.type.charAt(0));
  }

  public boolean isContentReferenced() {
    return isContentReferenced;
  }
}
