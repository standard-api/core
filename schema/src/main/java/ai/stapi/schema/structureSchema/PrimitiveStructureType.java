package ai.stapi.schema.structureSchema;

public class PrimitiveStructureType extends AbstractStructureType {

  public static final String SERIALIZATION_TYPE = "PrimitiveStructureType";
  public static final String KIND = AbstractStructureType.PRIMITIVE_TYPE;

  protected PrimitiveStructureType() {
    super();
  }

  protected PrimitiveStructureType(
      String serializationType,
      String definitionType,
      String description,
      boolean isAbstract,
      String parent
  ) {
    super(serializationType, AbstractStructureType.PRIMITIVE_TYPE, definitionType, description,
        isAbstract, parent);
  }

  public PrimitiveStructureType(
      String definitionType,
      String description,
      boolean isAbstract,
      String parent
  ) {
    super(SERIALIZATION_TYPE, AbstractStructureType.PRIMITIVE_TYPE, definitionType, description,
        isAbstract, parent);
  }
}
