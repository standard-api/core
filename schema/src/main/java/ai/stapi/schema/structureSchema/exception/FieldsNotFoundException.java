package ai.stapi.schema.structureSchema.exception;


import ai.stapi.schema.structureSchemaMapper.UnresolvableType;
import java.util.List;

public class FieldsNotFoundException extends RuntimeException {

  private FieldsNotFoundException(
      String message,
      String serializationType,
      List<String> fieldNames
  ) {
    super(
        String.format(
            "%s not found at structure with serializationType '%s', %s",
            fieldNames.size() == 1 ? String.format("Field '%s'", fieldNames.get(0)) :
                String.format("Fields '%s'", fieldNames),
            serializationType,
            message
        )
    );
  }

  private FieldsNotFoundException(
      String message,
      String serializationType
  ) {
    super(
        String.format(
            "Some fields not found at structure with serializationType '%s', %s",
            serializationType,
            message
        )
    );
  }

  public static FieldsNotFoundException becauseSerializationTypeDoesNotExistInSchema(
      String serializationType,
      List<String> fieldNames
  ) {
    return new FieldsNotFoundException(
        "because serializationType does not exist in schema.",
        serializationType,
        fieldNames
    );
  }

  public static FieldsNotFoundException becauseSerializationTypeIsPrimitive(
      String serializationType,
      List<String> fieldNames
  ) {
    return new FieldsNotFoundException(
        "because serializationType is primitive.",
        serializationType,
        fieldNames
    );
  }

  public static FieldsNotFoundException becauseSerializationTypeIsOfUnknownType(
      String serializationType
  ) {
    return new FieldsNotFoundException(
        "because serializationType '%s' is of unknown type.",
        serializationType
    );
  }

  public static FieldsNotFoundException becauseSerializationTypeDoesNotExistInSchema(
      String serializationType
  ) {
    return new FieldsNotFoundException(
        "because serializationType does not exist in schema.",
        serializationType
    );
  }

  public static FieldsNotFoundException becauseSerializationTypeDoesNotExistInSchema(
      String serializationType,
      UnresolvableType unresolvableType
  ) {
    return new FieldsNotFoundException(
        String.format(
            "because serializationType does not exist in schema, but is unresolvable.%nMissing dependencies: '%s'",
            unresolvableType.missingDependencies()
        ),
        serializationType
    );
  }

  public static FieldsNotFoundException becauseSerializationTypeIsPrimitive(
      String serializationType
  ) {
    return new FieldsNotFoundException(
        "because serializationType is primitive.",
        serializationType
    );
  }

  public static FieldsNotFoundException becauseSerializationTypeIsOfUnknownType(
      String serializationType,
      List<String> fieldNames
  ) {
    return new FieldsNotFoundException(
        "because serializationType '%s' is of unknown type.",
        serializationType,
        fieldNames
    );
  }

  public static FieldsNotFoundException becauseFinderIsRestricted(
      String serializationType,
      List<String> fieldNames
  ) {
    return new FieldsNotFoundException(
        "because Structure Schema Finder is restricted. You should use Test Case supporting Structure Schema.",
        serializationType,
        fieldNames
    );
  }

  public static FieldsNotFoundException becauseFinderIsRestricted(
      String serializationType
  ) {
    return new FieldsNotFoundException(
        "because Structure Schema Finder is restricted. You should use Test Case supporting Structure Schema.",
        serializationType
    );
  }

  public static FieldsNotFoundException becauseSomeFieldsAreMissing(
      String serializationType,
      List<String> fieldNames
  ) {
    return new FieldsNotFoundException(
        "because they are missing in found Complex Structure Type.",
        serializationType,
        fieldNames
    );
  }
}
