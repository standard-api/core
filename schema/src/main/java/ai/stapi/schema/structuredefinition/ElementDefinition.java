package ai.stapi.schema.structuredefinition;

import ai.stapi.serialization.SerializableObject;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.List;
import org.jetbrains.annotations.Nullable;

public class ElementDefinition implements SerializableObject {

  public static final String SERIALIZATION_TYPE = "ElementDefinition";
  
  private String path;

  @Nullable
  private Integer min;

  @Nullable
  private String max;

  @Nullable
  private String shortDescription;

  @Nullable
  private String definition;

  @Nullable
  private String comment;

  @Nullable
  private String contentReference;

  private List<ElementDefinitionType> type;

  protected ElementDefinition() {
  }

  public ElementDefinition(
      String path,
      List<ElementDefinitionType> type,
      @Nullable Integer min,
      @Nullable String max,
      @Nullable String shortDescription,
      @Nullable String definition,
      @Nullable String comment
  ) {
    this.path = path;
    this.type = type;
    this.min = min;
    this.max = max;
    this.shortDescription = shortDescription;
    this.definition = definition;
    this.comment = comment;
  }

  public ElementDefinition(
      String path,
      List<ElementDefinitionType> type,
      @Nullable Integer min,
      @Nullable String max,
      @Nullable String shortDescription,
      @Nullable String definition,
      @Nullable String comment,
      @Nullable String contentReference
  ) {
    this.path = path;
    this.type = type;
    this.min = min;
    this.max = max;
    this.shortDescription = shortDescription;
    this.definition = definition;
    this.comment = comment;
    this.contentReference = contentReference;
  }

  public String getPath() {
    return path;
  }

  @Nullable
  public Integer getMin() {
    return min;
  }

  @Nullable
  public String getMax() {
    return max;
  }

  @Nullable
  @JsonProperty("short")
  public String getShortDescription() {
    return shortDescription;
  }

  @Nullable
  public String getDefinition() {
    return definition;
  }

  @Nullable
  public String getComment() {
    return comment;
  }

  public List<ElementDefinitionType> getType() {
    return this.type;
  }

  @Nullable
  public String getContentReference() {
    return contentReference;
  }

  @Override
  public String toString() {
    return "ElementDefinition{" +
        "path='" + path + '\'' +
        ", min=" + min +
        ", max='" + max + '\'' +
        ", shortDescription='" + shortDescription + '\'' +
        ", definition='" + definition + '\'' +
        ", comment='" + comment + '\'' +
        ", type=" + type +
        '}';
  }

  @Override
  public String getSerializationType() {
    return SERIALIZATION_TYPE;
  }
}