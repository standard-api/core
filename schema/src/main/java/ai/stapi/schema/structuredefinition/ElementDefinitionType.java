package ai.stapi.schema.structuredefinition;

import java.util.ArrayList;
import java.util.List;

public class ElementDefinitionType {

  private String code;
  private List<String> targetProfile;

  protected ElementDefinitionType() {
  }

  public ElementDefinitionType(
      String code,
      List<String> targetProfile
  ) {
    this.code = code;
    this.targetProfile = targetProfile;
  }

  public ElementDefinitionType(String code) {
    this.code = code;
    this.targetProfile = new ArrayList<>();
  }

  public String getCode() {
    return code;
  }

  public List<String> getTargetProfile() {
    return targetProfile;
  }
}
