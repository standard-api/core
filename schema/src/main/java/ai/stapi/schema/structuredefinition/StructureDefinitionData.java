package ai.stapi.schema.structuredefinition;

import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.serialization.SerializableObject;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.List;

public class StructureDefinitionData implements SerializableObject {

  public static final String SERIALIZATION_TYPE = "StructureDefinition";

  private String id;
  private String url;
  private String status;
  private String description;
  private String kind;
  private Boolean isAbstract;
  private String type;
  private String baseDefinition;
  private UniqueIdentifier baseDefinitionReference;
  private Differential differential;

  protected StructureDefinitionData() {
  }

  public StructureDefinitionData(
      String id,
      String url,
      String status,
      String description,
      String kind,
      Boolean isAbstract,
      String type,
      String baseDefinition,
      UniqueIdentifier baseDefinitionReference,
      Differential differential
  ) {
    this.id = id;
    this.url = url;
    this.status = status;
    this.description = description;
    this.kind = kind;
    this.isAbstract = isAbstract;
    this.type = type;
    this.baseDefinition = baseDefinition;
    this.baseDefinitionReference = baseDefinitionReference;
    this.differential = differential;
  }

  public StructureDefinitionData(
      String id,
      String url,
      String status,
      String description,
      String kind,
      Boolean isAbstract,
      String type,
      String baseDefinition,
      UniqueIdentifier baseDefinitionReference,
      List<ElementDefinition> differential
  ) {
    this.id = id;
    this.url = url;
    this.status = status;
    this.description = description;
    this.kind = kind;
    this.isAbstract = isAbstract;
    this.type = type;
    this.baseDefinition = baseDefinition;
    this.baseDefinitionReference = baseDefinitionReference;
    this.differential = new Differential(differential, this.id);
  }

  public String getId() {
    return id;
  }

  public String getUrl() {
    return url;
  }

  public String getStatus() {
    return status;
  }

  public String getDescription() {
    return description;
  }

  public String getKind() {
    return kind;
  }

  @JsonProperty("abstract")
  public Boolean getIsAbstract() {
    return isAbstract;
  }

  public String getType() {
    return type;
  }

  public Differential getDifferential() {
    return differential;
  }

  public UniqueIdentifier getBaseDefinitionReference() {
    return baseDefinitionReference;
  }

  public String getBaseDefinition() {
    return baseDefinition;
  }

  @Override
  @JsonIgnore
  public String getSerializationType() {
    return StructureDefinitionData.SERIALIZATION_TYPE;
  }

  @Override
  public String toString() {
    return "StructureDefinitionDTO{" +
        "id='" + id + '\'' +
        ", url='" + url + '\'' +
        ", status='" + status + '\'' +
        ", description='" + description + '\'' +
        ", kind='" + kind + '\'' +
        ", isAbstract=" + isAbstract +
        ", type='" + type + '\'' +
        ", baseDefinition=" + baseDefinitionReference +
        ", differential=" + differential +
        '}';
  }

  public static class Differential implements SerializableObject {

    public static final String SERIALIZATION_TYPE = "StructureDefinitionDifferential";
    private List<ElementDefinition> element;
    private String parent;

    private Differential() {
    }

    public Differential(List<ElementDefinition> element, String parent) {
      this.element = element;
      this.parent = parent;
    }

    public List<ElementDefinition> getElement() {
      return element;
    }

    public String getParent() {
      return parent;
    }

    @Override
    public String toString() {
      return "Differential{" +
          "element=" + element +
          '}';
    }

    @Override
    public String getSerializationType() {
      return SERIALIZATION_TYPE;
    }
  }
}
