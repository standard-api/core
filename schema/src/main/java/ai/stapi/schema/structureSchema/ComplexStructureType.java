package ai.stapi.schema.structureSchema;

import ai.stapi.schema.structureSchema.exception.FieldDoesNotExist;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.stream.Collectors;

public class ComplexStructureType extends AbstractStructureType {

  public static final String SERIALIZATION_TYPE = "ComplexStructureType";
  public static final String KIND = AbstractStructureType.COMPLEX_TYPE;

  private Map<String, FieldDefinition> fields;

  protected ComplexStructureType() {
    super();
  }

  protected ComplexStructureType(
      String serializationType,
      String kind,
      String definitionType,
      Map<String, FieldDefinition> fields,
      String description,
      String parent,
      boolean isAbstract
  ) {
    super(serializationType, kind, definitionType, description, isAbstract, parent);
    this.fields = new HashMap<>(fields);
  }

  public ComplexStructureType(
      String definitionType,
      Map<String, FieldDefinition> fields,
      String description,
      String parent,
      boolean isAbstract

  ) {
    super(
        SERIALIZATION_TYPE,
        AbstractStructureType.COMPLEX_TYPE,
        definitionType,
        description,
        isAbstract,
        parent
    );
    this.fields = new HashMap<>(fields);
  }

  public Map<String, FieldDefinition> getAllFields() {
    return fields;
  }

  public Map<String, FieldDefinition> getOwnFields() {
    return fields.entrySet().stream()
        .filter(entry -> entry.getValue().getParentDefinitionType().equals(this.definitionType))
        .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
  }

  public FieldDefinition getField(String fieldName) {
    if (!this.fields.containsKey(fieldName)) {
      throw new FieldDoesNotExist(
          String.format(
              "Field '%s' does not exist on '%s'.",
              fieldName,
              this.getDefinitionType()
          )
      );
    }
    return this.fields.get(fieldName);
  }

  public boolean hasField(String fieldName) {
    return this.fields.containsKey(fieldName);
  }

  public ComplexStructureType copyWithNewFields(Map<String, FieldDefinition> newFields) {
    return new ComplexStructureType(
        definitionType,
        newFields,
        description,
        parent,
        isAbstract
    );
  }
}
