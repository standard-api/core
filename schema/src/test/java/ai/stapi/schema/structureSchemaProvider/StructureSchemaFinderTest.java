package ai.stapi.schema.structureSchemaProvider;

import ai.stapi.schema.structureSchema.ComplexStructureType;
import ai.stapi.schema.structureSchema.FieldDefinition;
import ai.stapi.schema.structureSchema.FieldType;
import ai.stapi.schema.structureSchema.PrimitiveStructureType;
import ai.stapi.schema.structureSchema.StructureSchema;
import ai.stapi.schema.structureSchema.exception.FieldsNotFoundException;
import ai.stapi.schema.structuredefinition.ElementDefinition;
import ai.stapi.schema.structuredefinition.ElementDefinitionType;
import ai.stapi.schema.structuredefinition.StructureDefinitionData;
import ai.stapi.schema.test.integration.IntegrationTestCase;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Import;

@Import(StubStructureSchemaProvider.class)
class StructureSchemaFinderTest extends IntegrationTestCase {

  @Autowired
  private StructureSchemaFinder structureSchemaFinder;

  @Autowired
  private StubStructureSchemaProvider structureSchemaProvider;

  @Test
  void itWillFailGettingFieldForNonExistingField() {
    this.structureSchemaProvider.setSchema(
        new StructureSchema(
            new PrimitiveStructureType(
                "primitive",
                "This is primitive",
                false,
                ""
            ),
            new ComplexStructureType(
                "Element",
                new HashMap<>(),
                "Required type for primitive",
                "",
                false
            )
        )
    );

    Executable throwable =
        () -> this.structureSchemaFinder.getFieldDefinitionFor("nonExistingType", "any");

    this.thenExceptionMessageApproved(
        FieldsNotFoundException.class,
        throwable
    );
  }

  @Test
  void itWillFailGettingFieldForPrimitive() {
    this.structureSchemaProvider.setSchema(
        new StructureSchema(
            new PrimitiveStructureType(
                "primitive",
                "This is primitive",
                false,
                ""
            ),
            new ComplexStructureType(
                "Element",
                new HashMap<>(),
                "Required type for primitive",
                "",
                false
            )
        )
    );

    Executable throwable =
        () -> this.structureSchemaFinder.getFieldDefinitionFor("primitive", "any");

    this.thenExceptionMessageApproved(
        FieldsNotFoundException.class,
        throwable
    );
  }

  @Test
  void itWillGetFieldDefinitionOnComplex() {

    this.structureSchemaProvider.setSchema(
        new StructureSchema(
            new ComplexStructureType(
                "Element",
                new HashMap<>(),
                "Required type for primitive",
                "",
                false
            ),
            new PrimitiveStructureType(
                "primitive",
                "This is primitive",
                false,
                ""
            ),
            new ComplexStructureType(
                "TestComplex",
                new HashMap<>(
                    Map.of(
                        "testedField",
                        new FieldDefinition(
                            "testedField",
                            0,
                            "*",
                            "this is parent's field",
                            new ArrayList<>(List.of(new FieldType("primitive", "primitive"))),
                            "TestComplex"
                        )
                    )
                ),
                "This is new complex type",
                "",
                false
            )
        )
    );

    this.thenObjectApproved(
        this.structureSchemaFinder.getFieldDefinitionFor("TestComplex", "testedField"));
    Assertions.assertTrue(this.structureSchemaFinder.isList("TestComplex", "testedField"));
  }

  @Test
  void itWillGetFieldDefinitionOnBoxed() {
    this.structureSchemaProvider.add(
        new ArrayList<>(List.of(
            new StructureDefinitionData(
                "testPrimitive",
                "http://test.url/for/primitive",
                "draft",
                "Structure Definition for test primitive",
                "primitive-type",
                false,
                "testPrimitive",
                null,
                null,
                new ArrayList<>()
            ),
            new StructureDefinitionData(
                "Element",
                "http://test.url/for/Element",
                "draft",
                "Structure Definition for Element",
                "complex-type",
                false,
                "Element",
                null,
                null,
                new ArrayList<>(
                    List.of(
                        new ElementDefinition(
                            "Element.exampleLeafField",
                            new ArrayList<>(List.of(new ElementDefinitionType("testPrimitive"))),
                            0,
                            "1",
                            "example field with primitive",
                            "",
                            ""
                        ),
                        new ElementDefinition(
                            "Element.exampleListField",
                            new ArrayList<>(List.of(new ElementDefinitionType("testPrimitive"))),
                            0,
                            "*",
                            "example field with primitive",
                            "",
                            ""
                        )
                    )
                )
            )
        ))
    );

    this.thenObjectApproved(
        this.structureSchemaFinder.getFieldDefinitionFor("BoxedTestPrimitive", "exampleLeafField"));
    Assertions.assertFalse(
        this.structureSchemaFinder.isList("BoxedTestPrimitive", "exampleLeafField"));
    Assertions.assertTrue(
        this.structureSchemaFinder.isList("BoxedTestPrimitive", "exampleListField"));
  }
}