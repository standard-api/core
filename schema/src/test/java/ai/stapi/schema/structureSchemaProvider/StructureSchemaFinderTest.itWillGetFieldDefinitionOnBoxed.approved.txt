{
  "serializationType" : "FieldDefinition",
  "description" : "",
  "max" : "1",
  "min" : 0,
  "name" : "exampleLeafField",
  "parentDefinitionType" : "Element",
  "types" : [ {
    "isContentReferenced" : false,
    "originalType" : "testPrimitive",
    "type" : "testPrimitive",
    "typeGroup" : "STANDARD"
  } ]
}